compiler-core/src/type_/expression.rs

```diff
@@ -1757,14 +1757,22 @@ impl<'a, 'b, 'c> ExprTyper<'a, 'b, 'c> {
                 };
             }
 
-            body_typer.infer(body)
-        })?;
+            (body_typer.hydrator.rigid_names(), body_typer.infer(body))
+        });
 
-        // Check that any return type type is accurate.
-        if let Some(return_type) = return_type {
-            self.unify(return_type, body.type_())
-                .map_err(|e| e.return_annotation_mismatch().into_error(body.location()))?;
-        }
+        let body = body_infer
+            .and_then(|typed_body| {
+                // Check that return type is accurate.
+                match return_type {
+                    Some(return_type) => self.unify(return_type, typed_body.type_()).map_err(|e| {
+                        e.return_annotation_mismatch()
+                            .into_error(typed_body.location())
+                    }),
+                    _ => Ok(()),
+                }
+                .map(|_| typed_body)
+            })
+            .map_err(|e| e.with_unify_error_rigid_names(body_rigid_names))?;
```

Comment:
I found this a bit hard to read, could we go back to using `if let Some`?

To make it easier the `in_new_scope` can be changed to return `Result<(_, _)>` rather than `(Result<_>, _)`, and then we can keep the `?` on the end of that line as before.

```rust
let body = body_typer.infer(body)?;
Ok((body_typer.hydrator.rigid_names(), body))
```

Reply:
I'll try to refactor this code to make it more readable.